File: examples/fortran/example_06_curvilinear_grid.f90
program example_06_curvilinear_grid
  use cmip7_fortran_common
  implicit none

  integer, parameter :: nx = 4
  integer, parameter :: ny = 3
  integer, parameter :: nvertices = 4
  character(len=1024) :: repo_root
  character(len=1024) :: output_dir
  character(len=2048) :: tables_path
  character(len=2048) :: input_path
  integer :: x_id
  integer :: y_id
  integer :: time_id
  integer :: grid_id
  integer :: grid_table_id
  integer :: var_id
  integer :: ierr
  integer :: nul_pos
  integer :: i
  integer :: j
  integer :: t
  real :: hfls(nx, ny, ntimes)
  character(len=2048) :: filename
  double precision :: x(nx)
  double precision :: y(ny)
  double precision :: x_bnds(2, nx)
  double precision :: y_bnds(2, ny)
  double precision :: latitude(nx, ny)
  double precision :: longitude(nx, ny)
  double precision :: latitude_vertices(nvertices, nx, ny)
  double precision :: longitude_vertices(nvertices, nx, ny)
  double precision :: time(ntimes)
  double precision :: time_bnds(2, ntimes)
  character(len=30) :: parameter_names(6)
  character(len=1) :: parameter_units(6)
  double precision :: parameter_values(6)

  call get_example_args(repo_root, output_dir)
  call prepare_cmor_paths(repo_root, output_dir, "example_06_input.json", &
       "mon", "r1", "f1", tables_path, input_path)

  ierr = cmor_setup(inpath=trim(tables_path), &
       netcdf_file_action=CMOR_REPLACE, &
       exit_control=CMOR_EXIT_ON_MAJOR)
  call check_status("cmor_setup", ierr)

  ierr = cmor_dataset_json(trim(input_path))
  call check_status("cmor_dataset_json", ierr)

  grid_table_id = cmor_load_table("CMIP7_grids.json")
  call cmor_set_table(grid_table_id)

  x = (/ 0.0d0, 10000.0d0, 20000.0d0, 30000.0d0 /)
  x_bnds(:, 1) = (/ -5000.0d0, 5000.0d0 /)
  x_bnds(:, 2) = (/ 5000.0d0, 15000.0d0 /)
  x_bnds(:, 3) = (/ 15000.0d0, 25000.0d0 /)
  x_bnds(:, 4) = (/ 25000.0d0, 35000.0d0 /)
  y = (/ 0.0d0, 10000.0d0, 20000.0d0 /)
  y_bnds(:, 1) = (/ -5000.0d0, 5000.0d0 /)
  y_bnds(:, 2) = (/ 5000.0d0, 15000.0d0 /)
  y_bnds(:, 3) = (/ 15000.0d0, 25000.0d0 /)

  x_id = cmor_axis(table_entry="x", units="m", length=nx, &
       coord_vals=x, cell_bounds=x_bnds)
  y_id = cmor_axis(table_entry="y", units="m", length=ny, &
       coord_vals=y, cell_bounds=y_bnds)

  do j = 1, ny
    do i = 1, nx
      latitude(i, j) = 10.0d0 * dble(j) - 2.0d0 * dble(i - 1)
      longitude(i, j) = 280.0d0 + 10.0d0 * dble(i - 1) &
           + 2.0d0 * dble(j - 1)
      latitude_vertices(:, i, j) = (/ latitude(i, j) - 5.0d0, &
           latitude(i, j) - 4.0d0, latitude(i, j) + 5.0d0, &
           latitude(i, j) + 4.0d0 /)
      longitude_vertices(:, i, j) = (/ longitude(i, j) - 5.0d0, &
           longitude(i, j) + 5.0d0, longitude(i, j) + 5.0d0, &
           longitude(i, j) - 5.0d0 /)
    enddo
  enddo

  grid_id = cmor_grid(axis_ids=(/ x_id, y_id /), &
       latitude=latitude, &
       longitude=longitude, &
       latitude_vertices=latitude_vertices, &
       longitude_vertices=longitude_vertices)

  parameter_names = (/ "standard_parallel1           ", &
       "longitude_of_central_meridian", &
       "latitude_of_projection_origin", &
       "false_easting                ", &
       "false_northing               ", &
       "standard_parallel2           " /)
  parameter_units = (/ " ", " ", " ", " ", " ", " " /)
  parameter_values = (/ -20.0d0, 175.0d0, 13.0d0, 8.0d0, 0.0d0, 20.0d0 /)
  ierr = cmor_set_grid_mapping(grid_id, "lambert_conformal_conic", &
       parameter_names, parameter_values, parameter_units)
  call check_status("cmor_set_grid_mapping", ierr)

  ierr = cmor_load_table("CMIP7_atmos.json")
  time = (/ 15.5d0, 45.5d0 /)
  time_bnds(:, 1) = (/ 0.0d0, 31.0d0 /)
  time_bnds(:, 2) = (/ 31.0d0, 60.0d0 /)
  time_id = cmor_axis(table="CMIP7_atmos.json", &
       table_entry="time", &
       units="days since 1979-01-01", &
       length=ntimes, &
       coord_vals=time, &
       cell_bounds=time_bnds)

  do t = 1, ntimes
    do j = 1, ny
      do i = 1, nx
        hfls(i, j, t) = 80.0 + 2.0 * real(i - 1) + 8.0 * real(j - 1) &
             + real(t - 1)
      enddo
    enddo
  enddo

  var_id = cmor_variable(table="CMIP7_atmos.json", &
       table_entry="hfls_tavg-u-hxy-u", &
       units="W m-2", &
       axis_ids=(/ grid_id, time_id /), &
       positive="up", &
       missing_value=missing_value)

  ierr = cmor_write(var_id, hfls)
  call check_status("cmor_write", ierr)

  filename = ""
  ierr = cmor_close(var_id, file_name=filename)
  call check_status("cmor_close(var)", ierr)
  nul_pos = index(filename, char(0))
  if (nul_pos > 0) filename(nul_pos:) = " "
  write(*, '(a)') trim(filename)

  ierr = cmor_close()
  call check_status("cmor_close", ierr)
end program example_06_curvilinear_grid

File: examples/c/example_06_curvilinear_grid.c
#include "cmip7_c_common.h"

#include <stdio.h>

#define NX 4
#define NY 3
#define NVERTICES 4
#define NPARAMS 6
#define PARAM_LEN 32
#define UNIT_LEN 2

int main(int argc, char **argv) {
  const char *repo_root = argc > 1 ? argv[1] : ".";
  const char *output_dir = argc > 2 ? argv[2] : "examples/c/output";
  char input_path[CMIP7_PATH_MAX];
  char tables_path[CMIP7_PATH_MAX];
  char cell_measures[CMOR_MAX_STRING];
  char long_name[CMOR_MAX_STRING];
  char filename[CMOR_MAX_STRING];
  int file_action = CMOR_REPLACE;
  int exit_control = CMOR_EXIT_ON_MAJOR;
  int table_id, grid_table_id, x_id, y_id, time_id, grid_id, var_id;
  int grid_axes[2];
  int axes[2];
  int i, j, t;
  double x[NX] = {0.0, 10000.0, 20000.0, 30000.0};
  double y[NY] = {0.0, 10000.0, 20000.0};
  double x_bnds[NX + 1] = {-5000.0, 5000.0, 15000.0, 25000.0, 35000.0};
  double y_bnds[NY + 1] = {-5000.0, 5000.0, 15000.0, 25000.0};
  double latitude[NY * NX];
  double longitude[NY * NX];
  double latitude_vertices[NY * NX * NVERTICES];
  double longitude_vertices[NY * NX * NVERTICES];
  double time[CMIP7_NTIMES] = {15.5, 45.5};
  double time_bnds[CMIP7_NTIMES + 1] = {0.0, 31.0, 60.0};
  char parameter_names[NPARAMS][PARAM_LEN] = {"standard_parallel1",
                                              "longitude_of_central_meridian",
                                              "latitude_of_projection_origin",
                                              "false_easting",
                                              "false_northing",
                                              "standard_parallel2"};
  char parameter_units[NPARAMS][UNIT_LEN] = {"", "", "", "", "", ""};
  double parameter_values[CMOR_MAX_GRID_ATTRIBUTES] = {-20.0, 175.0, 13.0,
                                                       8.0,   0.0,   20.0};
  float hfls[CMIP7_NTIMES * NY * NX];
  float missing = CMIP7_MISSING_VALUE;

  snprintf(tables_path, sizeof(tables_path), "%s/cmip7-cmor-tables/tables",
           repo_root);
  cmip7_write_user_input_json(output_dir, "example_06_input.json", "mon", "r1",
                              "f1", input_path, sizeof(input_path));

  cmor_setup(tables_path, &file_action, NULL, &exit_control, NULL, NULL);
  cmor_dataset_json(input_path);

  cmor_load_table("CMIP7_grids.json", &grid_table_id);
  cmor_set_table(grid_table_id);
  cmor_axis(&y_id, "y", "m", NY, y, 'd', y_bnds, 1, NULL);
  cmor_axis(&x_id, "x", "m", NX, x, 'd', x_bnds, 1, NULL);

  for (j = 0; j < NY; ++j) {
    for (i = 0; i < NX; ++i) {
      int idx = j * NX + i;
      int vertex_idx = idx * NVERTICES;

      latitude[idx] = 10.0 * (double)(j + 1) - 2.0 * (double)i;
      longitude[idx] = 280.0 + 10.0 * (double)i + 2.0 * (double)j;
      latitude_vertices[vertex_idx + 0] = latitude[idx] - 5.0;
      latitude_vertices[vertex_idx + 1] = latitude[idx] - 4.0;
      latitude_vertices[vertex_idx + 2] = latitude[idx] + 5.0;
      latitude_vertices[vertex_idx + 3] = latitude[idx] + 4.0;
      longitude_vertices[vertex_idx + 0] = longitude[idx] - 5.0;
      longitude_vertices[vertex_idx + 1] = longitude[idx] + 5.0;
      longitude_vertices[vertex_idx + 2] = longitude[idx] + 5.0;
      longitude_vertices[vertex_idx + 3] = longitude[idx] - 5.0;
    }
  }

  grid_axes[0] = y_id;
  grid_axes[1] = x_id;
  cmor_grid(&grid_id, 2, grid_axes, 'd', latitude, longitude, NVERTICES,
            latitude_vertices, longitude_vertices);
  cmor_set_grid_mapping(grid_id, "lambert_conformal_conic", NPARAMS,
                        &parameter_names[0][0], PARAM_LEN, parameter_values,
                        &parameter_units[0][0], UNIT_LEN);

  cmor_load_table("CMIP7_atmos.json", &table_id);
  cmor_axis(&time_id, "time", "days since 1979-01-01", CMIP7_NTIMES, time, 'd',
            time_bnds, 1, NULL);

  for (t = 0; t < CMIP7_NTIMES; ++t) {
    for (j = 0; j < NY; ++j) {
      for (i = 0; i < NX; ++i) {
        int idx = (t * NY + j) * NX + i;
        hfls[idx] = 80.0f + 2.0f * (float)i + 8.0f * (float)j + (float)t;
      }
    }
  }

  axes[0] = time_id;
  axes[1] = grid_id;
  cmor_variable(&var_id, "hfls_tavg-u-hxy-u", "W m-2", 2, axes, 'f', &missing,
                NULL, "up", NULL, NULL, NULL);

  cmip7_get_cell_measures(tables_path, "atmos", "hfls_tavg-u-hxy-u", "mon",
                          "glb", cell_measures, sizeof(cell_measures));
  cmor_set_variable_attribute(var_id, "cell_measures", 'c', cell_measures);
  if (cmip7_get_long_name_override(tables_path, "atmos", "hfls_tavg-u-hxy-u",
                                   "mon", "glb", long_name,
                                   sizeof(long_name))) {
    cmor_set_variable_attribute(var_id, "long_name", 'c', long_name);
  }

  cmor_write(var_id, hfls, 'f', NULL, CMIP7_NTIMES, NULL, NULL, NULL);
  filename[0] = '\0';
  cmor_close_variable(var_id, filename, NULL);
  printf("%s\n", filename);
  cmor_close();
  return 0;
}
